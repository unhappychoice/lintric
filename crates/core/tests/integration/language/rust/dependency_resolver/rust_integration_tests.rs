use insta::assert_snapshot;
use lintric_core::analyze_code;
use std::fs;

macro_rules! test_dependency_resolver_analysis {
    ($test_name:ident, $fixture_name:literal, $ir_snapshot:literal) => {
        #[test]
        fn $test_name() {
            let file_path = format!(
                "tests/integration/language/rust/dependency_resolver/fixtures/{}.rs",
                $fixture_name
            );
            let (ir, _result) = analyze_code(file_path.clone()).unwrap();

            let source_code = fs::read_to_string(&file_path).unwrap();
            let ast_result = lintric_core::get_s_expression_from_content(
                source_code.clone(),
                lintric_core::Language::Rust,
            );
            let ast_string = ast_result.unwrap_or_else(|e| format!("AST Error: {}", e));

            let ir_snapshot_content = format!(
                "Source Code:\n{}\n\nAST:\n{}\n\nIR:\n{:#?}",
                source_code, ast_string, ir
            );

            assert_snapshot!($ir_snapshot, ir_snapshot_content);

            // Verify dependencies are detected
            assert!(!ir.dependencies.is_empty(), "Should have dependencies");
        }
    };
}

test_dependency_resolver_analysis!(
    test_basic_symbol_resolution,
    "basic_symbol_resolution",
    "dependency_resolver_basic_symbol_resolution_ir"
);

test_dependency_resolver_analysis!(
    test_method_resolution,
    "method_resolution",
    "dependency_resolver_method_resolution_ir"
);

test_dependency_resolver_analysis!(
    test_generic_resolution,
    "generic_resolution",
    "dependency_resolver_generic_resolution_ir"
);

test_dependency_resolver_analysis!(
    test_complex_trait_hierarchy,
    "complex_trait_hierarchy",
    "dependency_resolver_complex_trait_hierarchy_ir"
);

test_dependency_resolver_analysis!(
    test_scope_resolution,
    "scope_resolution",
    "dependency_resolver_scope_resolution_ir"
);

test_dependency_resolver_analysis!(
    test_module_resolution,
    "module_resolution",
    "dependency_resolver_module_resolution_ir"
);

test_dependency_resolver_analysis!(
    test_forward_references,
    "forward_references",
    "dependency_resolver_forward_references_ir"
);

test_dependency_resolver_analysis!(
    test_closure_dependencies,
    "closure_dependencies",
    "dependency_resolver_closure_dependencies_ir"
);

test_dependency_resolver_analysis!(
    test_advanced_generics,
    "advanced_generics",
    "dependency_resolver_advanced_generics_ir"
);

test_dependency_resolver_analysis!(
    test_macro_dependencies,
    "macro_dependencies",
    "dependency_resolver_macro_dependencies_ir"
);

test_dependency_resolver_analysis!(
    test_async_dependencies,
    "async_dependencies",
    "dependency_resolver_async_dependencies_ir"
);

test_dependency_resolver_analysis!(
    test_pattern_matching,
    "pattern_matching",
    "dependency_resolver_pattern_matching_ir"
);

#[test]
fn test_specific_dependency_types() {
    let file_path =
        "tests/integration/language/rust/dependency_resolver/fixtures/method_resolution.rs";
    let (ir, _result) = analyze_code(file_path.to_string()).unwrap();

    // Check that we have some dependencies
    assert!(!ir.dependencies.is_empty(), "Should have some dependencies");
}

#[test]
fn test_trait_method_dependencies() {
    let file_path =
        "tests/integration/language/rust/dependency_resolver/fixtures/complex_trait_hierarchy.rs";
    let (ir, _result) = analyze_code(file_path.to_string()).unwrap();

    // Verify trait method dependencies are resolved
    let trait_methods: Vec<_> = ir
        .dependencies
        .iter()
        .filter(|d| d.symbol == "name" || d.symbol == "speak" || d.symbol == "fur_color")
        .collect();

    assert!(
        !trait_methods.is_empty(),
        "Should resolve trait method dependencies"
    );
}

#[test]
fn test_generic_type_dependencies() {
    let file_path =
        "tests/integration/language/rust/dependency_resolver/fixtures/generic_resolution.rs";
    let (ir, _result) = analyze_code(file_path.to_string()).unwrap();

    // Check for generic-related dependencies
    let generic_deps: Vec<_> = ir
        .dependencies
        .iter()
        .filter(|d| d.symbol.contains("Container") || d.symbol == "process")
        .collect();

    assert!(
        !generic_deps.is_empty(),
        "Should resolve generic type dependencies"
    );
}
