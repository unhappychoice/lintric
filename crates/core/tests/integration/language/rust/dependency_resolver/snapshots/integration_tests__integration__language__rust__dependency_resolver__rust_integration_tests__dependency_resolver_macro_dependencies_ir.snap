---
source: crates/core/tests/integration/language/rust/dependency_resolver/rust_integration_tests.rs
expression: ir_snapshot_content
---
Source Code:
macro_rules! create_function {
    ($name:ident, $return_type:ty) => {
        fn $name() -> $return_type {
            Default::default()
        }
    };
}

macro_rules! impl_display {
    ($type:ty) => {
        impl std::fmt::Display for $type {
            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(f, "Custom Display")
            }
        }
    };
}

create_function!(get_number, i32);
create_function!(get_string, String);

struct CustomType;

impl_display!(CustomType);

fn main() {
    let num = get_number();
    let s = get_string();
    let custom = CustomType;
    
    println!("{} {} {}", num, s, custom);
}

AST:
(source_file
  (macro_definition
    name: (identifier "create_function")
    (macro_rule
      left: (token_tree_pattern
        (token_binding_pattern
          name: (metavariable)
          type: (fragment_specifier)
        )
        (token_binding_pattern
          name: (metavariable)
          type: (fragment_specifier)
        )
      )
      right: (token_tree
        (metavariable)
        (token_tree)
        (metavariable)
        (token_tree
          (identifier "Default")
          (token_tree)
        )
      )
    )
  )
  (macro_definition
    name: (identifier "impl_display")
    (macro_rule
      left: (token_tree_pattern
        (token_binding_pattern
          name: (metavariable)
          type: (fragment_specifier)
        )
      )
      right: (token_tree
        (identifier "std")
        (identifier "fmt")
        (identifier "Display")
        (metavariable)
        (token_tree
          (identifier "fmt")
          (token_tree
            (self "self")
            (identifier "f")
            (mutable_specifier)
            (identifier "std")
            (identifier "fmt")
            (identifier "Formatter")
          )
          (identifier "std")
          (identifier "fmt")
          (identifier "Result")
          (token_tree
            (identifier "write")
            (token_tree
              (identifier "f")
              (string_literal ""Custom Display""
                (string_content)
              )
            )
          )
        )
      )
    )
  )
  (expression_statement
    (macro_invocation
      macro: (identifier "create_function")
      (token_tree
        (identifier "get_number")
        (primitive_type "i32")
      )
    )
  )
  (expression_statement
    (macro_invocation
      macro: (identifier "create_function")
      (token_tree
        (identifier "get_string")
        (identifier "String")
      )
    )
  )
  (struct_item
    name: (type_identifier "CustomType")
  )
  (expression_statement
    (macro_invocation
      macro: (identifier "impl_display")
      (token_tree
        (identifier "CustomType")
      )
    )
  )
  (function_item
    name: (identifier "main")
    parameters: (parameters)
    body: (block
      (let_declaration
        pattern: (identifier "num")
        value: (call_expression
          function: (identifier "get_number")
          arguments: (arguments)
        )
      )
      (let_declaration
        pattern: (identifier "s")
        value: (call_expression
          function: (identifier "get_string")
          arguments: (arguments)
        )
      )
      (let_declaration
        pattern: (identifier "custom")
        value: (identifier "CustomType")
      )
      (expression_statement
        (macro_invocation
          macro: (identifier "println")
          (token_tree
            (string_literal ""{} {} {}""
              (string_content)
            )
            (identifier "num")
            (identifier "s")
            (identifier "custom")
          )
        )
      )
    )
  )
)

IR:
IntermediateRepresentation {
    file_path: "tests/integration/language/rust/dependency_resolver/fixtures/macro_dependencies.rs",
    definitions: [
        Definition {
            name: "create_function",
            position: Position {
                start_line: 1,
                start_column: 14,
                end_line: 1,
                end_column: 29,
            },
            definition_type: MacroDefinition,
        },
        Definition {
            name: "$name",
            position: Position {
                start_line: 2,
                start_column: 6,
                end_line: 2,
                end_column: 11,
            },
            definition_type: MacroVariableDefinition,
        },
        Definition {
            name: "$return_type",
            position: Position {
                start_line: 2,
                start_column: 19,
                end_line: 2,
                end_column: 31,
            },
            definition_type: MacroVariableDefinition,
        },
        Definition {
            name: "impl_display",
            position: Position {
                start_line: 9,
                start_column: 14,
                end_line: 9,
                end_column: 26,
            },
            definition_type: MacroDefinition,
        },
        Definition {
            name: "$type",
            position: Position {
                start_line: 10,
                start_column: 6,
                end_line: 10,
                end_column: 11,
            },
            definition_type: MacroVariableDefinition,
        },
        Definition {
            name: "CustomType",
            position: Position {
                start_line: 22,
                start_column: 8,
                end_line: 22,
                end_column: 18,
            },
            definition_type: StructDefinition,
        },
        Definition {
            name: "main",
            position: Position {
                start_line: 26,
                start_column: 4,
                end_line: 26,
                end_column: 8,
            },
            definition_type: FunctionDefinition,
        },
        Definition {
            name: "num",
            position: Position {
                start_line: 27,
                start_column: 9,
                end_line: 27,
                end_column: 12,
            },
            definition_type: VariableDefinition,
        },
        Definition {
            name: "s",
            position: Position {
                start_line: 28,
                start_column: 9,
                end_line: 28,
                end_column: 10,
            },
            definition_type: VariableDefinition,
        },
        Definition {
            name: "custom",
            position: Position {
                start_line: 29,
                start_column: 9,
                end_line: 29,
                end_column: 15,
            },
            definition_type: VariableDefinition,
        },
    ],
    dependencies: [
        Dependency {
            source_line: 3,
            target_line: 2,
            symbol: "$name",
            dependency_type: MacroVariable,
            context: Some(
                "Metavariable:3:12",
            ),
        },
        Dependency {
            source_line: 3,
            target_line: 2,
            symbol: "$return_type",
            dependency_type: MacroVariable,
            context: Some(
                "Metavariable:3:23",
            ),
        },
        Dependency {
            source_line: 11,
            target_line: 10,
            symbol: "$type",
            dependency_type: MacroVariable,
            context: Some(
                "Metavariable:11:36",
            ),
        },
        Dependency {
            source_line: 19,
            target_line: 1,
            symbol: "create_function",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:19:1",
            ),
        },
        Dependency {
            source_line: 20,
            target_line: 1,
            symbol: "create_function",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:20:1",
            ),
        },
        Dependency {
            source_line: 24,
            target_line: 9,
            symbol: "impl_display",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:24:1",
            ),
        },
        Dependency {
            source_line: 24,
            target_line: 22,
            symbol: "CustomType",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:24:15",
            ),
        },
        Dependency {
            source_line: 29,
            target_line: 22,
            symbol: "CustomType",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:29:18",
            ),
        },
        Dependency {
            source_line: 31,
            target_line: 27,
            symbol: "num",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:31:26",
            ),
        },
        Dependency {
            source_line: 31,
            target_line: 28,
            symbol: "s",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:31:31",
            ),
        },
        Dependency {
            source_line: 31,
            target_line: 29,
            symbol: "custom",
            dependency_type: VariableUse,
            context: Some(
                "Identifier:31:34",
            ),
        },
    ],
    usage: [
        Usage {
            name: "create_function",
            kind: Identifier,
            position: Position {
                start_line: 1,
                start_column: 14,
                end_line: 1,
                end_column: 29,
            },
            context: None,
        },
        Usage {
            name: "$name",
            kind: Metavariable,
            position: Position {
                start_line: 2,
                start_column: 6,
                end_line: 2,
                end_column: 11,
            },
            context: None,
        },
        Usage {
            name: "$return_type",
            kind: Metavariable,
            position: Position {
                start_line: 2,
                start_column: 19,
                end_line: 2,
                end_column: 31,
            },
            context: None,
        },
        Usage {
            name: "$name",
            kind: Metavariable,
            position: Position {
                start_line: 3,
                start_column: 12,
                end_line: 3,
                end_column: 17,
            },
            context: None,
        },
        Usage {
            name: "$return_type",
            kind: Metavariable,
            position: Position {
                start_line: 3,
                start_column: 23,
                end_line: 3,
                end_column: 35,
            },
            context: None,
        },
        Usage {
            name: "Default",
            kind: Identifier,
            position: Position {
                start_line: 4,
                start_column: 13,
                end_line: 4,
                end_column: 20,
            },
            context: None,
        },
        Usage {
            name: "impl_display",
            kind: Identifier,
            position: Position {
                start_line: 9,
                start_column: 14,
                end_line: 9,
                end_column: 26,
            },
            context: None,
        },
        Usage {
            name: "$type",
            kind: Metavariable,
            position: Position {
                start_line: 10,
                start_column: 6,
                end_line: 10,
                end_column: 11,
            },
            context: None,
        },
        Usage {
            name: "std",
            kind: Identifier,
            position: Position {
                start_line: 11,
                start_column: 14,
                end_line: 11,
                end_column: 17,
            },
            context: None,
        },
        Usage {
            name: "fmt",
            kind: Identifier,
            position: Position {
                start_line: 11,
                start_column: 19,
                end_line: 11,
                end_column: 22,
            },
            context: None,
        },
        Usage {
            name: "Display",
            kind: Identifier,
            position: Position {
                start_line: 11,
                start_column: 24,
                end_line: 11,
                end_column: 31,
            },
            context: None,
        },
        Usage {
            name: "$type",
            kind: Metavariable,
            position: Position {
                start_line: 11,
                start_column: 36,
                end_line: 11,
                end_column: 41,
            },
            context: None,
        },
        Usage {
            name: "fmt",
            kind: Identifier,
            position: Position {
                start_line: 12,
                start_column: 16,
                end_line: 12,
                end_column: 19,
            },
            context: None,
        },
        Usage {
            name: "f",
            kind: Identifier,
            position: Position {
                start_line: 12,
                start_column: 27,
                end_line: 12,
                end_column: 28,
            },
            context: None,
        },
        Usage {
            name: "std",
            kind: Identifier,
            position: Position {
                start_line: 12,
                start_column: 35,
                end_line: 12,
                end_column: 38,
            },
            context: None,
        },
        Usage {
            name: "fmt",
            kind: Identifier,
            position: Position {
                start_line: 12,
                start_column: 40,
                end_line: 12,
                end_column: 43,
            },
            context: None,
        },
        Usage {
            name: "Formatter",
            kind: Identifier,
            position: Position {
                start_line: 12,
                start_column: 45,
                end_line: 12,
                end_column: 54,
            },
            context: None,
        },
        Usage {
            name: "std",
            kind: Identifier,
            position: Position {
                start_line: 12,
                start_column: 59,
                end_line: 12,
                end_column: 62,
            },
            context: None,
        },
        Usage {
            name: "fmt",
            kind: Identifier,
            position: Position {
                start_line: 12,
                start_column: 64,
                end_line: 12,
                end_column: 67,
            },
            context: None,
        },
        Usage {
            name: "Result",
            kind: Identifier,
            position: Position {
                start_line: 12,
                start_column: 69,
                end_line: 12,
                end_column: 75,
            },
            context: None,
        },
        Usage {
            name: "write",
            kind: Identifier,
            position: Position {
                start_line: 13,
                start_column: 17,
                end_line: 13,
                end_column: 22,
            },
            context: None,
        },
        Usage {
            name: "f",
            kind: Identifier,
            position: Position {
                start_line: 13,
                start_column: 24,
                end_line: 13,
                end_column: 25,
            },
            context: None,
        },
        Usage {
            name: "create_function",
            kind: Identifier,
            position: Position {
                start_line: 19,
                start_column: 1,
                end_line: 19,
                end_column: 16,
            },
            context: None,
        },
        Usage {
            name: "get_number",
            kind: Identifier,
            position: Position {
                start_line: 19,
                start_column: 18,
                end_line: 19,
                end_column: 28,
            },
            context: None,
        },
        Usage {
            name: "create_function",
            kind: Identifier,
            position: Position {
                start_line: 20,
                start_column: 1,
                end_line: 20,
                end_column: 16,
            },
            context: None,
        },
        Usage {
            name: "get_string",
            kind: Identifier,
            position: Position {
                start_line: 20,
                start_column: 18,
                end_line: 20,
                end_column: 28,
            },
            context: None,
        },
        Usage {
            name: "String",
            kind: Identifier,
            position: Position {
                start_line: 20,
                start_column: 30,
                end_line: 20,
                end_column: 36,
            },
            context: None,
        },
        Usage {
            name: "impl_display",
            kind: Identifier,
            position: Position {
                start_line: 24,
                start_column: 1,
                end_line: 24,
                end_column: 13,
            },
            context: None,
        },
        Usage {
            name: "CustomType",
            kind: Identifier,
            position: Position {
                start_line: 24,
                start_column: 15,
                end_line: 24,
                end_column: 25,
            },
            context: None,
        },
        Usage {
            name: "get_number",
            kind: CallExpression,
            position: Position {
                start_line: 27,
                start_column: 15,
                end_line: 27,
                end_column: 27,
            },
            context: Some(
                "call_expression",
            ),
        },
        Usage {
            name: "get_string",
            kind: CallExpression,
            position: Position {
                start_line: 28,
                start_column: 13,
                end_line: 28,
                end_column: 25,
            },
            context: Some(
                "call_expression",
            ),
        },
        Usage {
            name: "CustomType",
            kind: Identifier,
            position: Position {
                start_line: 29,
                start_column: 18,
                end_line: 29,
                end_column: 28,
            },
            context: None,
        },
        Usage {
            name: "println",
            kind: Identifier,
            position: Position {
                start_line: 31,
                start_column: 5,
                end_line: 31,
                end_column: 12,
            },
            context: None,
        },
        Usage {
            name: "num",
            kind: Identifier,
            position: Position {
                start_line: 31,
                start_column: 26,
                end_line: 31,
                end_column: 29,
            },
            context: None,
        },
        Usage {
            name: "s",
            kind: Identifier,
            position: Position {
                start_line: 31,
                start_column: 31,
                end_line: 31,
                end_column: 32,
            },
            context: None,
        },
        Usage {
            name: "custom",
            kind: Identifier,
            position: Position {
                start_line: 31,
                start_column: 34,
                end_line: 31,
                end_column: 40,
            },
            context: None,
        },
    ],
    analysis_metadata: AnalysisMetadata {
        language: "Rust",
        total_lines: 32,
        analysis_timestamp: "now",
        lintric_version: "0.1.0",
    },
}
